### PR TITLE
glib: implement FromGlibContainer for GString

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Needed by https://github.com/gtk-rs/gtk4-rs/pull/89

### CODE REVIEW

File: glib/src/gstring.rs [hunk 1]
```diff
@@ -404,6 +404,48 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
     }
 }
 
+#[doc(hidden)]
+impl<'a> FromGlibContainer<*const c_char, *const i8> for GString {
+    unsafe fn from_glib_none_num(ptr: *const i8, num: usize) -> Self {
```
Comment (bilelmoussaoui) [inline glib/src/gstring.rs:409]: To be honest, I'm not very sure if this was the right approach at all :) I'm a bit confused what would be the differences between the 3 methods in such case?

Comment (sdroege) [reply glib/src/gstring.rs:409]: `none` doesn't take ownership of the input pointer, the other two would (and are the same because `char` is a basic type without memory management).

File: glib/src/gstring.rs [hunk 2]
```diff
@@ -404,6 +404,48 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
...
+        let mut bytes = Vec::with_capacity(num);
+
+        for i in 0..num {
+            let b = std::ptr::read(ptr.add(i)) as u8;
+            bytes.push(b);
+        }
+        CStr::from_bytes_with_nul_unchecked(bytes.as_slice()).into()
+    }
+
+    unsafe fn from_glib_container_num(ptr: *const i8, num: usize) -> Self {
```
Comment (sdroege) [inline glib/src/gstring.rs:422]: `container` and `full` could simply create a `GString(Inner::Foreign(ptr as *mut _, num))` without copying or anything

File: glib/src/gstring.rs [hunk 3]
```diff
@@ -404,6 +404,48 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
...
+        if num == 0 {
+            return Self::from("");
+        }
+        let mut bytes = Vec::with_capacity(num);
+
+        for i in 0..num {
+            let b = std::ptr::read(ptr.add(i)) as u8;
+            bytes.push(b);
+        }
+        CStr::from_bytes_with_nul_unchecked(bytes.as_slice()).into()
```
Comment (sdroege) [inline glib/src/gstring.rs:419]: Does this actually work? I believe that function requires the vec to be NUL terminated or otherwise it's UB. You could add a 0 byte in the end though

File: glib/src/gstring.rs [hunk 4]
```diff
@@ -404,6 +404,48 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
...
+    unsafe fn from_glib_none_num(ptr: *const i8, num: usize) -> Self {
+        if num == 0 {
+            return Self::from("");
+        }
+        let mut bytes = Vec::with_capacity(num);
+
+        for i in 0..num {
+            let b = std::ptr::read(ptr.add(i)) as u8;
+            bytes.push(b);
+        }
```
Comment (sdroege) [inline glib/src/gstring.rs:418]:
```rust
let slice = std::slice::from_raw_parts(ptr as *const u8, num);
bytes.extend_from_slice(slice);
```

That does the same as your loop but more efficiently and shorter.

Comment (bilelmoussaoui) [reply glib/src/gstring.rs:418]: Ah I somehow got confused by the docs of slice::from_raw_parts where it says that num shouldn't be the number of bytes but rather the number of elements.

Comment (sdroege) [reply glib/src/gstring.rs:418]: That's the same for the `FromGlibContainer` trait btw (and for `u8` / `i8` it's all the same anyway).

File: glib/src/gstring.rs (hunk 3 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:418]: Should be working now

File: glib/src/gstring.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:421]: I've added a very simple test just to check if everything was working fine but the `container` & `full` one fails with a `free(): invalid pointer`, I wonder whether what I wrote is just completely wrong on that test or something isn't working properly? :)

File: glib/src/gstring.rs (hunk 1 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:409]: Noted, thanks!

File: glib/src/gstring.rs [hunk 5]
```diff
@@ -404,6 +404,29 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
     }
 }
 
+#[doc(hidden)]
+impl<'a> FromGlibContainer<*const c_char, *const i8> for GString {
```
Comment (sdroege) [inline glib/src/gstring.rs:408]: For completeness, can you also implement it for `*mut _` and `u8`?

File: glib/src/gstring.rs [hunk 6]
```diff
@@ -404,6 +404,29 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
...
+    unsafe fn from_glib_none_num(ptr: *const i8, num: usize) -> Self {
+        if num == 0 {
+            return Self::from("");
+        }
+        let mut bytes = Vec::with_capacity(num);
```
Comment (sdroege) [inline glib/src/gstring.rs:413]:
```suggestion
        let mut bytes = Vec::with_capacity(num + 1);
```

For the 0 :)

File: glib/src/gstring.rs [hunk 7]
```diff
@@ -511,6 +534,21 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_from_glib_container() {
+        unsafe {
+            let a: GString =
+                FromGlibContainer::from_glib_container_num("hello_world".as_ptr() as *const i8, 5);
+            assert_eq!("hello", a.as_str());
+            let b: GString =
+                FromGlibContainer::from_glib_full_num("hello_world".as_ptr() as *const i8, 5);
```
Comment (sdroege) [inline glib/src/gstring.rs:544]:
```suggestion
                FromGlibContainer::from_glib_full_num(glib_sys::g_strdup("hello_world".as_ptr() as *const i8), 5);
```

Or similar. The problem is that you pass a static (and definitely not heap allocated with the GLib allocator) string in there but say that `GString` can take ownership of it. That doesn't work :)

Same for the container variant as they're really the same here.

File: glib/src/gstring.rs (hunk 5 above)
Comment (sdroege) [reply glib/src/gstring.rs:408]: You can implement that all by calling into the other implementation by casting the raw pointers.

File: glib/src/gstring.rs (hunk 2 above)
Comment (sdroege) [reply glib/src/gstring.rs:421]: See comments below

File: glib/src/gstring.rs (hunk 6 above)
Comment (sdroege) [reply glib/src/gstring.rs:413]: Otherwise you'd allocate `num` bytes, copy the string, append `0` and that appending would allocate another memory area on the heap of at least `num+1` bytes, copy everything first, then free the old memory, then append the `0`.

File: glib/src/gstring.rs (hunk 5 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:408]: > You can implement that all by calling into the other implementation by casting the raw pointers.

I was going to copy paste them :P

File: glib/src/gstring.rs (hunk 7 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs]: I should have looked at the other tests first..

File: glib/src/gstring.rs (hunk 6 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs]: Interesting, thanks as always!

File: glib/src/gstring.rs [hunk 8]
```diff
@@ -404,6 +404,74 @@ impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
...
+
+        CStr::from_bytes_with_nul_unchecked(bytes.as_slice()).into()
+    }
+
+    unsafe fn from_glib_container_num(ptr: *const i8, num: usize) -> Self {
+        GString(Inner::Foreign(ptr as *mut _, num))
+    }
+
+    unsafe fn from_glib_full_num(ptr: *const i8, num: usize) -> Self {
+        GString(Inner::Foreign(ptr as *mut _, num))
```
Comment (sdroege) [inline glib/src/gstring.rs:426]: I wonder if these work correctly for empty strings or some fixes would be needed. Can you add a test for empty strings in the different variants?

Comment (sdroege) [reply glib/src/gstring.rs:426]: I.e. `NULL` / `0` and `$random_pointer` / `0` should give the same results for `none` and `full`

Comment (bilelmoussaoui) [reply glib/src/gstring.rs:426]: I have added more cases to the test

File: glib/src/gstring.rs [hunk 9]
```diff
@@ -511,6 +579,32 @@ mod tests {
...
+
+            let c: GString =
+                FromGlibContainer::from_glib_none_num(std::ptr::null() as *const i8, 0);
+            assert_eq!("", c.as_str());
+
+            let d: GString = FromGlibContainer::from_glib_none_num("".as_ptr(), 0);
+            assert_eq!("", d.as_str());
+
+            let e: GString = FromGlibContainer::from_glib_container_num(
+                ffi::g_strdup(std::ptr::null() as *const i8),
```
Comment (sdroege) [inline glib/src/gstring.rs:601]: `g_strdup()` returns `NULL` if passed `NULL` or not? And the cast should be unneeded

File: glib/src/gstring.rs [hunk 10]
```diff
@@ -511,6 +579,32 @@ mod tests {
...
+                0,
```
Comment (sdroege) [inline glib/src/gstring.rs:602]: And I think this actually is UB because it calls `slice::from_raw_parts()` somewhere with `NULL`. See https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html

The `as_str()` and other functions further above would need to special-case `NULL` / `0` and simply return a `""` in that case.

File: glib/src/gstring.rs (hunk 9 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:601]: Yes, per the function documentation 
> Duplicates a string. If str is NULL it returns NULL. The returned string should be freed with g_free() when no longer needed.

The cast is needed because otherwise I keep getting 
```
 FromGlibContainer::from_glib_none_num(std::ptr::null(), 0);
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for struct `gstring::GString`
```

File: glib/src/gstring.rs (hunk 10 above)
Comment (bilelmoussaoui) [reply glib/src/gstring.rs:602]: Does that has to happen on all other functions? otherwise I think we can just add a check to `full` & `container` implementation, the same way I did for the `none` one

File: glib/src/gstring.rs (hunk 9 above)
Comment (sdroege) [reply glib/src/gstring.rs:601]: `std::ptr::null::<i8>()` might be nicer than a cast then :)

Comment (bilelmoussaoui) [reply glib/src/gstring.rs:601]: Aha. Thanks! I wasn't aware I can do this :)

File: glib/src/gstring.rs (hunk 10 above)
Comment (sdroege) [reply glib/src/gstring.rs:608]: Well, `GString::from_glib_full(ptr::null())` seems to be allowed and would end up in exactly the same situation (before your changes already).

Probably best to handle that generally somewhere.

Comment (bilelmoussaoui) [reply glib/src/gstring.rs:607]: It should be handled now

File: glib/src/gstring.rs [hunk 11]
```diff
@@ -62,13 +61,15 @@ impl GString {
...
     /// length, underneath the `GString` instance for the duration of the borrow.
     unsafe fn new_borrowed(ptr: *const c_char) -> Borrowed<Self> {
-        assert!(!ptr.is_null());
         Borrowed::new(GString(Inner::Foreign(ptr as *mut _, libc::strlen(ptr))))
     }
 
     pub fn as_str(&self) -> &str {
         let cstr = match self {
             GString(Inner::Foreign(ptr, length)) => unsafe {
+                if ptr.is_null() || length == &0 {
```
Comment (sdroege) [inline glib/src/gstring.rs:70]: Why `&0` instead of `0`? :)

Comment (bilelmoussaoui) [reply glib/src/gstring.rs:70]: Because length is &usize and there's no PartialEq impl for that

Comment (sdroege) [reply glib/src/gstring.rs:70]: Ok!

**Review by sdroege — APPROVED** [review]
